core/species.cpp

```diff
@@ -3854,126 +4026,405 @@ void Species::SimplifyTreeSequence(void)
+	// this code is copied from https://github.com/tskit-dev/tskit/pull/2665/files (multichrom_wright_fisher.c)
 	{
-#if 0
-		// If we didn't keep first-generation lineages, or remember haplosomes, >1 root would mean not coalesced
-		if (tsk_tree_get_num_roots(&t) > 1)
+		tsk_size_t sample_count = (tsk_size_t)samples.size();
+		int ret;
+		const tsk_size_t num_nodes = main_tables.nodes.num_rows;
+		tsk_bool_t *keep_nodes = (tsk_bool_t *)calloc(num_nodes, sizeof(tsk_bool_t));	// note: cleared by calloc
+		tsk_id_t *node_id_map = (tsk_id_t *)malloc(num_nodes * sizeof(tsk_id_t));
+		
+		if (!keep_nodes || !node_id_map)
+			EIDOS_TERMINATION << "ERROR (Species::SimplifyAllTreeSequences): allocation failed; you may need to raise the memory limit for SLiM." << EidosTerminate(nullptr);
+		
+		// mark the nodes we want to keep: samples (including remembered nodes), plus all nodes referenced by edges
+		for (tsk_size_t j = 0; j < sample_count; j++)
+			keep_nodes[samples[j]] = true;
+		
+		for (Chromosome *chromosome : chromosomes_)
 		{
-			fully_coalesced = false;
-			break;
+			slim_chromosome_index_t chromosome_index = chromosome->Index();
+			TreeSeqInfo &chromosome_tsinfo = treeseq_[chromosome_index];
+			tsk_table_collection_t &chromosome_tables = chromosome_tsinfo.tables_;
+			tsk_id_t *edge_child = chromosome_tables.edges.child;
+			tsk_id_t *edge_parent = chromosome_tables.edges.parent;
+			tsk_size_t edges_num_rows = chromosome_tables.edges.num_rows;
+			
+			for (tsk_size_t k = 0; k < edges_num_rows; k++) {
+				keep_nodes[edge_child[k]] = true;
+				keep_nodes[edge_parent[k]] = true;
+			}
 		}
-#else
-		// But we do have retained/remembered nodes in the tree, so we need to be smarter; nodes for the first gen
-		// ancestors will always be present, giving >1 root in each tree even when we have coalesced, and the
-		// remembered individuals may mean that more than one root node has children, too, even when we have
-		// coalesced.  What we need to know is: how many roots are there that have >0 *extant* children?  This
-		// is what we use the tracked samples for; they are extant individuals.
-		for (tsk_id_t root = tsk_tree_get_left_root(&t); root != TSK_NULL; root = t.right_sib[root])
+		
+		// tskit does the work for us and provides an index map
+		ret = tsk_node_table_keep_rows(&main_tables.nodes, keep_nodes, 0, node_id_map);
+		if (ret < 0) handle_error("tsk_node_table_keep_rows", ret);
+		
+		// remap node references
+		for (Chromosome *chromosome : chromosomes_)
 		{
-			int64_t num_tracked = t.num_tracked_samples[root];
+			slim_chromosome_index_t chromosome_index = chromosome->Index();
+			TreeSeqInfo &chromosome_tsinfo = treeseq_[chromosome_index];
+			tsk_table_collection_t &chromosome_tables = chromosome_tsinfo.tables_;
+			tsk_id_t *edge_child = chromosome_tables.edges.child;
+			tsk_id_t *edge_parent = chromosome_tables.edges.parent;
+			tsk_size_t edges_num_rows = chromosome_tables.edges.num_rows;
```

Comment:
We also need to do `chromosome_tables.mutations.node`. I think we don't need to loop over `mutations.node` to check for which nodes to keep because `simplify` will have removed any mutations not ancestral to any samples; so the only mutations that we will still have that aren't on an edge are above either a root (which is parent to an edge, so will still be there) or above a sample (also still there). But we probably ought to check if `mutations_node[k]` is out of bounds for `node_id_map` and if so throw an internal error.

Reply:
Perhaps you could magnify on this comment a bit.  I haven't delved into the code again yet, but I suspect I'm going to need clarification here when I do.  :->

Reply:
The `mutation` table has a `node` column; here we are changing which node is which, so we need to also remap them, like `mutations_node[k] = node_id_map[mutations_node[k]]`.

You might also think we need to loop through the mutation table to add nodes that are referred to there to `keep_nodes`, but I don't think that's true - we should be able to `assert node_id_map[mutations_node[k]] >= 0`. (it'll be `-1` if the node has been removed)

Reply:
Jerome's example didn't have mutations.

Reply:
oh p.s. also here edge*s*_child, etcetera

Reply:
OK, that last comment – what about edges_child?  And what's the etcetera?  You know this stuff way better than I do; if there are other references I need to be fixing up, please enumerate them exhaustively if you can!

Reply:
sorry, I meant replace `edge_child` with `edges_child`, as below

Reply:
OK, figured out what was meant by this, and I think I did it right.  :->  Resolving.